### PR TITLE
fix(security): redact email PII from auth-flow logs (#4254)

### LINKED ISSUE #4254: fix(security): redact email PII from auth-flow logs
## Background

Reviewer @rabble flagged a PII-in-log concern on PR #3784 ([review thread](https://github.com/divinevideo/divine-mobile/pull/3784#issuecomment-4419836344) point 4):

> **PII in warning log.** `PendingVerificationService.load()` logs the email address when discarding pre-fix pending verification. Safer to omit or redact the email unless auth email logging is already explicitly accepted.

A repo-wide audit found this is the established convention across the auth flow — **7 call sites log email at info/warning level** today. Redacting only the one line rabble flagged would create inconsistency without security benefit; the systematic fix was deferred from #3784's emergency-security scope to its own follow-up.

This issue tracks the systematic redaction.

## Sites to redact (audit, on `origin/main`)

| File | Line | Level | Current shape |
|---|---|---|---|
| `lib/blocs/divine_auth/divine_auth_cubit.dart` | 275 | Log.info | `'Email verification required for $email'` |
| `lib/blocs/divine_auth/divine_auth_cubit.dart` | 388 | Log.info | `'Sending password reset email to $email'` |
| `lib/blocs/email_verification/email_verification_cubit.dart` | 75 | Log.info | `'startPolling called for $email ...'` |
| `lib/services/pending_verification_service.dart` | 68 | Log.info | `'Saved pending verification for $email'` |
| `lib/services/pending_verification_service.dart` | 139 | Log.info | `'Pending verification for $email has expired, clearing'` |
| `lib/services/pending_verification_service.dart` | 148 | Log.info | `'Loaded pending verification for $email'` |

Plus one site on the unmerged #3784 branch:

| File | Line | Level | Current shape |
|---|---|---|---|
| `lib/services/pending_verification_service.dart` | 118 (post-#3784) | Log.warning | `'Discarding pre-fix pending verification for $email ...'` |

The Crashlytics path through `Reportable.toString()` strips `npub`/`nsec` but not email — broadening the sanitizer is part of this issue.

## Open decision: redaction strategy

| Option | Output for `user@example.com` | Trade-off |
|---|---|---|
| **A — opaque** | `[REDACTED]` | Matches existing `redactedSensitiveLogPlaceholder` convention; loses triage value |
| **B — partial (recommended)** | `u***@example.com` | Preserves domain for debugging; lets ops correlate same-account log entries; GDPR-friendly |

## Affected files

| File | Action | Description |
|---|---|---|
| `lib/utils/sensitive_uri_for_logs.dart` | Modify | Add `redactEmailForLogs(String)` helper |
| `lib/observability/reportable_error.dart` | Modify | Extend `sanitizeForCrashReport` to strip email patterns |
| `lib/blocs/divine_auth/divine_auth_cubit.dart` | Modify | Redact 2 call sites (275, 388) |
| `lib/blocs/email_verification/email_verification_cubit.dart` | Modify | Redact 1 call site (75) |
| `lib/services/pending_verification_service.dart` | Modify | Redact 3 call sites (68, 139, 148) on `main`; +1 site (118) after #3784 merges |
| `test/utils/sensitive_uri_for_logs_test.dart` | Modify (or Create) | Unit tests for the new helper |
| `test/observability/reportable_error_test.dart` | Modify (or Create) | Tests for email stripping in Crashlytics path |

## Implementation steps (bottom-up)

1. **Util — add `redactEmailForLogs`** in `lib/utils/sensitive_uri_for_logs.dart`. Partial-redactor (Option B) that returns `redactedSensitiveLogPlaceholder` for empty / malformed input. Single regex pass, no allocation on empty input.
2. **Observability — broaden Crashlytics sanitizer** in `lib/observability/reportable_error.dart`. Add an `_emailPattern` regex parallel to `_npubPattern` / `_nsecPattern`; replace matches with the partial-redacted form. Defense-in-depth even if a call site forgets to redact.
3. **BLoCs — redact 3 call sites**: `divine_auth_cubit.dart:275`, `divine_auth_cubit.dart:388`, `email_verification_cubit.dart:75`. Pattern: `'… for ${redactEmailForLogs(email)}'`.
4. **Service — redact 3 call sites** in `pending_verification_service.dart` (68, 139, 148). Same pattern.
5. **Tests** — new helper-level tests + Crashlytics sanitizer tests. Verify existing behavioural tests in the 7-site files still pass.

## Testing

| Layer | Test File | What to test |
|---|---|---|
| Util | `test/utils/sensitive_uri_for_logs_test.dart` | `redactEmailForLogs` across valid, malformed, unicode, single-char local-part |
| Observability | `test/observability/reportable_error_test.dart` | `sanitizeForCrashReport` strips email alongside npub/nsec |
| Regression | existing tests in `pending_verification_service_test.dart`, `email_verification_cubit_test.dart`, `divine_auth_cubit_test.dart` | Confirm behaviour-level tests still pass |

## Risks

- **Log correlation loss** — ops may rely on email-in-logs for incident triage. Mitigation: Option B preserves enough to correlate.
- **Missed future sites** — new code may add email logs that bypass the helper. Mitigation: rely on review; optional follow-up to add a CI guard script (mirror of `grep_for_nsec_payload.sh`).
- **Scope creep** — other PII (phone, names) may exist in logs. Out of scope for this issue; file separately if discovered.

## Coordination with #3784

This issue's PR is independent of #3784 — they touch overlapping code in `pending_verification_service.dart` but neither blocks the other. Whichever PR merges second receives a small rebase to also redact line 118 (the migration-warning site #3784 added).

### PR DESCRIPTION
**Pull request by realmeylisdev**
## Description

Systematic redaction of email PII from auth-flow logs. Flagged in #3784's review ([point 4](https://github.com/divinevideo/divine-mobile/pull/3784#issuecomment-4419836344)) but kept out of the emergency-security scope of that PR; this PR is the dedicated follow-up.

Audit of `origin/main` found **7 sites** that log email at info/warning level across `PendingVerificationService`, `DivineAuthCubit`, `EmailVerificationCubit`, and `EmailVerificationScreen`. Redacting only the one line Rabble flagged would create inconsistency without security benefit; this PR addresses all 7 sites via a single helper plus a broadened Crashlytics sanitizer.

**Related Issue:** Closes #4254

## What's in this PR

| File | Change |
|------|--------|
| `mobile/lib/utils/sensitive_uri_for_logs.dart` | New `redactEmailForLogs(String)` helper (Option B — partial-redact, preserves domain) |
| `mobile/lib/observability/reportable_error.dart` | Broaden `sanitizeForCrashReport` to strip emails alongside npub/nsec |
| `mobile/lib/services/pending_verification_service.dart` | Wrap 3 log call sites (save / expired / loaded) |
| `mobile/lib/blocs/divine_auth/divine_auth_cubit.dart` | Wrap 2 log call sites (verification-required / password-reset) |
| `mobile/lib/blocs/email_verification/email_verification_cubit.dart` | Wrap 1 log call site (startPolling) |
| `mobile/lib/screens/auth/email_verification_screen.dart` | Wrap 1 log call site (persisted auto-login path) |
| `mobile/test/unit/utils/sensitive_uri_for_logs_test.dart` | +9 cases for the new helper |
| `mobile/test/observability/reportable_error_test.dart` | +4 cases for email stripping in the Crashlytics path |
| `mobile/test/screens/auth/email_verification_screen_test.dart` | Regression test for persisted auto-login log redaction |

## Redaction strategy

Chose **Option B — partial** over **Option A — opaque** (`[REDACTED]`):

```text
user@example.com   → u***@example.com
alice@gmail.com    → a***@gmail.com
first.last+tag@x.y → f***@x.y
```

Rationale: preserves the domain so ops can spot patterns like "all gmail.com users are failing" without identifying individual accounts. Local-part collapses to a fixed-width `x***` mask regardless of length so original length isn't leaked.

Empty / malformed input (no `@`, empty local-part, domain without `.`) falls back to the existing `redactedSensitiveLogPlaceholder` (`[REDACTED]`).

## Type of Change

- [x] 🛠️ Bug fix (non-breaking change which fixes an issue)

## Coordination with #3784

#3784 (still open at time of writing) adds one overlapping auth-log site on top of the 7 handled here:

- `pending_verification_service.dart:118` (post-#3784) — the legacy-nsec migration warning.

Whichever PR merges second will need a tiny rebase to wrap that site too. There is no functional dependency between the two PRs; they only overlap in the same service.

## Test plan

- [x] `flutter analyze lib test integration_test` — no issues
- [x] `flutter test test/unit/utils/sensitive_uri_for_logs_test.dart` — 20/20 (9 new + 11 existing)
- [x] `flutter test test/observability/reportable_error_test.dart` — 17/17 (4 new + 13 existing)
- [x] `flutter test test/blocs/divine_auth/ test/blocs/email_verification/ test/services/email_verification_listener_test.dart` — 96/96 (regression on the originally wrapped auth-log sites)
- [x] `flutter test test/screens/auth/email_verification_screen_test.dart` — 16/16 (regression on the persisted auto-login log path)
- [x] Manual: verified the helper output for the documented shapes in unit tests

## Follow-ups (not in this PR)

- Optional CI guard mirroring `grep_for_nsec_payload.sh` to flag any new `Log\.(info|debug|warning|error).*email` patterns that bypass the helper.
- Other PII (phone numbers, full names) in logs — out of scope for this issue; file separately if discovered.

### CODE REVIEW

**Review by hm21 — APPROVED** [review]
LGTM

File: mobile/test/observability/reportable_error_test.dart [hunk 1]
```diff
@@ -111,5 +112,58 @@ void main() {
...
+        contains('@gmail.com'),
+      );
+    });
+
+    test('uses the shared email redactor so masking stays aligned', () {
+      const email = 'first.last+tag@example.com';
+
+      expect(
+        // This locks the crash-report fallback to the same mask shape as logs.
+        sanitizeForCrashReport('auth failed for $email'),
```
Comment (NotThatKindOfDrLiz) [inline mobile/test/observability/reportable_error_test.dart:164]: This assertion is the useful guardrail for the cleanup: if either redaction path changes shape later, the test fails instead of silently letting the two masking rules diverge.

File: mobile/lib/observability/reportable_error.dart [hunk 2]
```diff
@@ -56,17 +58,32 @@ final class Reportable<T extends Object> implements ReportableError {
...
+///   lockstep.
 String sanitizeForCrashReport(String input) {
   return input
       .replaceAll(_npubPattern, 'npub1<redacted>')
-      .replaceAll(_nsecPattern, 'nsec1<redacted>');
+      .replaceAll(_nsecPattern, 'nsec1<redacted>')
+      .replaceAllMapped(
+        _emailPattern,
+        // Reuse the log helper here so both redaction surfaces stay aligned.
+        (match) => redactEmailForLogs(match.group(0)!),
```
Comment (NotThatKindOfDrLiz) [inline mobile/lib/observability/reportable_error.dart:87]: Centralizing the email mask through `redactEmailForLogs()` removes a second security-sensitive implementation. That keeps log redaction and the Crashlytics fallback from drifting on edge cases.

**Review by NotThatKindOfDrLiz — APPROVED** [review]
Shared email redaction path is in place, the alignment regression test is covering the cleanup, and checks are green on the current head.

File: mobile/lib/observability/reportable_error.dart (hunk 2 above)
Comment (realmeylisdev) [reply mobile/lib/observability/reportable_error.dart:87]: > Centralizing the email mask through `redactEmailForLogs()` removes a second security-sensitive implementation. That keeps log redaction and the Crashlytics fallback from drifting on edge cases.

Agreed — routing the Crashlytics path through `redactEmailForLogs` was the right call. One masking rule, one helper to audit if the policy ever needs to change (e.g. opaque `[REDACTED]` instead of partial). The doc-comment now points future readers at the helper, so there's no temptation to reimplement the mask inline.

File: mobile/test/observability/reportable_error_test.dart (hunk 1 above)
Comment (realmeylisdev) [reply mobile/test/observability/reportable_error_test.dart:164]: > This assertion is the useful guardrail for the cleanup: if either redaction path changes shape later, the test fails instead of silently letting the two masking rules diverge.

Agreed — pinning `sanitizeForCrashReport(x)` to `redactEmailForLogs(x)` rather than to a literal expected string means the test moves with the helper. If we ever switch to opaque redaction or change the mask width, this test fails loudly and forces both surfaces to be updated together instead of one silently drifting.

**Comment by realmeylisdev** [comment]
Thanks @NotThatKindOfDrLiz and @hm21 — both review threads addressed and resolved. The Crashlytics path now routes through the shared `redactEmailForLogs` helper, and the test pins `sanitizeForCrashReport` to that helper rather than a literal mask, so future changes to either surface fail loudly instead of drifting silently.
